### ISSUE TITLE
IfcOpenBot

### ISSUE THREAD

**Issue by aothms** [IfcOpenShell/IfcOpenShell#1162]
Currently automated builds are being generated on a Mac Mini 2014 with 4gb RAM using VMs for Win and Linux builds. This is no longer a workable solution as the host and guests are fighting for the little amount of RAM and both are unresponsive. It has always been tight, but maybe it's a Win10 update that pushed it over the edge.

Options:
- Upgrade hardware and keep current approach
- Use CI/CD (see #1087 for azure)
- Use Mac server hardware (https://mspoweruser.com/amazon-ec2-mac-instances-mac-mini-hardware/)

**Comment by aothms** [comment]
cc @Moult you might want to switch to an older build for BlenderBIM given #1121

**Comment by Moult** [comment]
So since I have no expertise in knowing the right way forward, which do you think @aothms (and @tpaviot for the Azure solution)?

Please choose the best technical solution :) If funding is an obstacle I'm sure we can raise funds.

**Comment by tpaviot** [comment]
Azure is a solution that perfectly suits my needs for CI/CD. You can have a look at the last ifopenshell build at https://dev.azure.com/tpaviot/ifcopenshell/_build/results?buildId=715&view=results
Windows build is still buggy but it's just at the end of the process, while trying to import the python module (library linking issue). Mac and Linux host agents take around 30mn to complete the build, whereas windows build is 4 times slower.

@aothms For the CD part, you could add "Deliver a conda package" to the options list (I use to automatically upload the condapackage to the anaconda repository). And why not a "Deliver a nightly build conda package", indeed I noticed from 1 to 10 daily commits, that could be a good way to quickly release bugfixes or new features. @Moult I don't know the solution you chose for BlenderBim installation, honestly I did not try it, but it could be possible to automate the installation using conda directly from Python.

Let me know if you need any help, I use to walk the azure/ci/conda path and I'm aware of many (but not all) traps and dead ends.

**Comment by Moult** [comment]
@tpaviot the BlenderBIM Add-on package is a non-issue. It is pure Python and depends on the bot builds for the C++ stuff :)

**Comment by tpaviot** [comment]
@Moult I was talking about the way you deliver BlenderBim to the end user. On your website, I see a "Download" menu item and the a button that points to a .zip archive. After downloading I guess I have to extract the content to some specific location on my machine, right ?  These steps could be avoided and you could download/install the addon directly from Blender, by adding an "install BlenderBim plugin", in the same way you install plugins for SublimeText for example. A conda package could make this possible, and it could also help to install updates without requiring the user to remove old installation/download another release/extract to some location etc.

Note: I rebased the "azure-ci" branch of my ifcopenshell fork on top of the upstrea master branch Build is on the way, if you want to compare results with the gh bot https://dev.azure.com/tpaviot/ifcopenshell/_build/results?buildId=743&view=results

**Comment by Moult** [comment]
@tpaviot ah, actually that is how Blender works. In Blender there is an install button which you need to press and browse to the ZIP file. It's a little odd, but a ZIP file is the recommended way to distribute add-ons and this process is considered normal in Blender. The user doesn't extract anything - the in-Blender installer knows how to handle the ZIP file.

**Comment by tpaviot** [comment]
@Moult this looks pretty simple, I will try ;)

**Comment by Moult** [comment]
@tpaviot https://blenderbim.org/install-blenderbim.html

**Comment by Moult** [comment]
Curious on what happened to this? :)

**Comment by aothms** [comment]
In the end just rolled my own stuff using ssh and ansible on a dedicated cloud mac and AWS for win and linux. It'll get more robust over time and then I'll set it on auto pilot again.

**Comment by Krande** [comment]
Hi, I was going to propose github actions as an alternative, but I see from #1306 that this is already in the works :) I might see if I can make a windows-build equivalent using github actions, but so far I am struggling a bit with using the VS developer cmd to work in github actions.

@Moult Slightly related to this, I did experiment with a end-to-end pipeline for a blender addon that contains a self-updating feature [here](https://github.com/Krande/blender_addons) that uses github actions\release and github REST api to check for updates. You could take a look and see if it contains something useful that could work with your plans for blenderbim?

**Comment by aothms** [comment]
hold off on this for a while, I have a reasonable working ansible playbook now, I just need to clean up the repo (remove secrets) and then I can put it online.

**Comment by Moult** [comment]
@Krande cool trick! That would definitely be useful for the BlenderBIM Add-on. Would you be interested in building the feature?

**Comment by Krande** [comment]
@Moult Sure, I can give it a go! I probably need some help with the blenderbim compile steps though.

**Comment by aothms** [comment]
https://github.com/IfcOpenShell/IfcOpenBot exists now, Krande you have write access

**Comment by Moult** [comment]
@Krande there is a Makefile in `src/ifcblenderexport`. Just do `make dist PLATFORM=linux`. Other platform options are `win` and `macos`. No compilation required. It just downloads and bundles binaries.

**Comment by Krande** [comment]
So I made a test-release for blenderbim on linux on my fork here: 

https://github.com/Krande/IfcOpenShell/releases/tag/blender28-bim-210212-linux

I will try to create a multiplatform build when I find some time :)

BTW! I had to make a small edit to the makefile:

https://github.com/Krande/IfcOpenShell/blob/57b6f7df16b04f82a49ac1732a73bbc7fff952ed/src/ifcblenderexport/Makefile#L108

Basically change pystache path from `build/lib/pystache` to `build/lib.linux-x86_64-2.7/pystache`. Is that just me?

**Comment by Moult** [comment]
@Krande sorry for the late reply. That looks odd, is the "2.7" a reference to Python version 2.7 by any chance? If so, that probably needs further investigation.

This is great! Can you let me know how you created this? I have never used Github Actions before. It would be great to have more frequent builds!

**Comment by Krande** [comment]
Ah, yes. How embarrassing :) I forgot to change the github actions default python version (which by default is py27). I fixed it now (also changed to ubuntu as a build platform for all three versions) and it is producing all three variants of the blenderbim addon (win, linux and macOS) upon push to branch (I tested the windows release and it worked fine). 

We can also define unique commit phrases for whenever you wish to trigger publish release and\or create separate "stable"\"latest" release pipelines. 

Also. The current pipeline produces 3 separate releases (1 for each platform), but maybe adding the three platform variants as different assets to the same release might make more sense (and less cluttered). 

Sure, let me know how and when you'd like a quick rundown of how this works?
